src/main/java/io/supertokens/inmemorydb/Start.java

```diff
@@ -1619,4 +1627,126 @@ public void addInfoToNonAuthRecipesBasedOnUserId(String className, String userId
+        try {
+            int updatedCount = TOTPQueries.markDeviceAsVerified(this, userId, deviceName);
+            if (updatedCount == 0) {
+                TOTPDevice[] devices = TOTPQueries.getDevices(this, userId);
+                for (TOTPDevice device : devices) {
+                    if (device.deviceName.equals(deviceName) && device.verified) {
+                        return true; // Device was already verified
+                    }
+                }
+                // Device was not found:
+                throw new UnknownDeviceException();
+            }
+            return false; // Device was marked as verified
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public void deleteDevice(String userId, String deviceName)
+            throws StorageQueryException, UnknownDeviceException {
+        try {
+            int deletedCount = TOTPQueries.deleteDevice(this, userId, deviceName);
+            if (deletedCount == 0) {
+                throw new UnknownDeviceException();
+            }
+
+            // Note: This step is only required for in-memory databases.
+            // They don't have cascading deletes, so we need to manually delete the codes
+            TOTPQueries.removeUsedCodes(this, userId, deviceName);
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public void updateDeviceName(String userId, String oldDeviceName, String newDeviceName)
+            throws StorageQueryException, DeviceAlreadyExistsException,
+            UnknownDeviceException {
+        try {
+            int updatedCount = TOTPQueries.updateDeviceName(this, userId, oldDeviceName, newDeviceName);
+            if (updatedCount == 0) {
+                throw new UnknownDeviceException();
+            }
+        } catch (SQLException e) {
+            if (e.getMessage()
+                    .equals("[SQLITE_CONSTRAINT]  Abort due to constraint violation (UNIQUE constraint failed: "
+                            + Config.getConfig(this).getTotpUserDevicesTable() + ".user_id, "
+                            + Config.getConfig(this).getTotpUserDevicesTable() + ".device_name" + ")")) {
+                throw new DeviceAlreadyExistsException();
+            }
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public TOTPDevice[] getDevices(String userId)
+            throws StorageQueryException {
+        try {
+            return TOTPQueries.getDevices(this, userId);
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public boolean insertUsedCode(TOTPUsedCode usedCodeObj)
+            throws StorageQueryException, TotpNotEnabledException {
+        try {
+            TOTPDevice[] devices = TOTPQueries.getDevices(this, usedCodeObj.userId);
```

Comment:
add comment that this is only for in memory db.

Reply:
Transferred this logic to TOTP Queries using transactions..